### ISSUE TITLE
Incorrect type narrowing for generic property return value

### ISSUE THREAD

**Issue by kamalfarahani** [astral-sh/ty#2831]
### Summary

When narrowing a generic base class to a specific subclass using `isinstance()`, the type checker fails to properly infer the type of a property that returns the generic type parameter. Instead of inferring the correct type (`str` in the example), it falls back to `object`.

**Code example**
```python
class BaseContainer[T]: ...

class ContainerA[T](BaseContainer[T]):
    def __init__(self, value: T) -> None:
        self._value: T = value

    @property
    def value(self) -> T:
        return self._value

class ContainerB[T](BaseContainer[T]): ...

def f() -> BaseContainer[str]:
    return ContainerA("TY")

def main():
    x = f()
    if not isinstance(x, ContainerA):
        return

    y = x.value  # Type checker incorrectly infers y as object (should be str)
    c = y.capitalize()  # False type error: "object" has no attribute "capitalize"
```

**Expected behavior**
After the `isinstance(x, ContainerA)` check, the type checker should understand that:
1. `x` is of type `ContainerA[str]` (not just `ContainerA`)
2. Therefore `x.value` should be of type `str`
3. String methods like `capitalize()` should be available

**Actual behavior**
The type checker infers `y` as type `object`, causing a false positive error when trying to call string methods on it.

**Environment**
- `ty` version: 0.0.17

### Version

0.0.17

**Comment by sharkdp** [comment]
Thank you for your question. You might want to make sure that `BaseContainer` is not bivariant in `T` (it currently does not use `T` at all), and then read [this FAQ entry](https://docs.astral.sh/ty/reference/typing-faq/#what-is-toplistunknown-and-why-does-it-appear). Let us know if questions remain!

**Comment by kamalfarahani** [comment]
@sharkdp 
Thanks for your reply
I think this is a bug The original implementation was roughly:
```python
from typing import Generic, TypeVar

T = TypeVar("T", covariant=True)


class Maybe(Generic[T]): ...


class Nothing(Maybe[T]): ...


class Just(Maybe[T]):
    def __init__(self, value: T) -> None:
        self._value = value

    @property
    def value(self) -> T:
        return self._value
```

I'm using inheritance here to emulate sum types, that's why `Maybe` and `Nothing` don't contain a value of `T` type in them.
`pyright` in these cases infers types correctly.

**Comment by sharkdp** [comment]
Let's use the code from your second post and consider:
```py
def main(m: Maybe[str]):
    if not isinstance(m, Just):
        return

    reveal_type(m)  # ty: Maybe[str] & Just[object]
```
Here, the `isinstance` check narrows the type `Maybe[str]` by intersecting with the top-materialization of `Just[T]`, which is `Just[object]`, since `Just` is covariant in `T`. This accounts for the possibility that there might be classes that inherit from both `Maybe` and `Just` *with a different specialization than `str`* (that would also pass this `isinstance` check). So far, this is intended behavior, even if it differs from other type checkers.

To prevent this situation, we typically suggest that you can make one of your classes `@final`, but marking `Just` as `@final` here, does not help. ty still infers `Maybe[str] & Just[object]`. I think you're right though that this could be simplified to `Just[str]` if `Just` is final.

**Comment by kamalfarahani** [comment]
Thanks for explanation, before posting the issue I tried the `@final` solution and as you said problem still persisted.

> This accounts for the possibility that there might be classes that inherit from both Maybe and Just with a different specialization than str (that would also pass this isinstance check). So far, this is intended behavior, even if it differs from other type checkers.

I think this case can only happen if the third class inherits  `Just` like:
```python
class Dummy[T](Just[Any]):
    def __init__(self) -> None:
        self._value = None
```
Or not specifying type var for `Just` which I think is the same as above. In other cases if we set a concrete type var for `Just` when inheriting it this will not happen, right?

**Comment by sharkdp** [comment]
> In other cases if we set a concrete type var for Just when inheriting it this will not happen, right?

What exactly?

The following class would pass the `isinstance(…, Just)` check, right? And so it seems correct to reveal `Maybe[str] & Just[object]` if `Just` is not final.
```py
class C(Just[object], Maybe[str]): ...
```

**Comment by kamalfarahani** [comment]
Yea I see 
Thanks for answer

**Comment by carljm** [comment]
Thanks @kamalfarahani for the report and @sharkdp for analysis!

With the addition of `@final` on `Just`, our failure to simplify `Maybe[str] & Just[object]` to `Just[str]` is a bug (or missing feature in our intersection simplification), which is already tracked in #512.

Without `@final` on `Just`, I think our behavior here is as-designed, though we have #1578 open to track whether we should be willing to make more unsound assumptions in order to give the results people expect.
